app/controllers/visit_requests_controller.rb

```diff
@@ -31,7 +31,7 @@ def event
   end
 
   def visit_request
-    @visit_request ||= VisitRequest.find(params[:id])
+    @visit_request ||= VisitRequest.find_by!(token: params[:token])
```

Comment:
You don't need to do it here. Move it to `VisitRequest::FinalConfirmation` service. 
```ruby
VisitRequest::FinalConfirmation.call(params)
```

Reply:
Agree, that controller isn't the best place for this code, even though it is shared among the actions. Anyway, it was already placed here and it was the fastest way to achieve the goal.

Reply:
Sorry, my bad (misunderstood). Same for next comment.